File: 2013-06-07/python/exercise2.py
from pyplasm import *
from numpy import *

dom1D = INTERVALS(50)(32)
domain = PROD([dom1D,dom1D])

altitude = {};

def mounts(dom):
	x = dom[0];
	y = dom[1];
	z = random.random();
 	return [x, y, z];

terrain = MAP(mounts)(domain);

domainLake1d = INTERVALS(20)(32)
domainLake = PROD([domainLake1d,domainLake1d])


z_lake = 0;
z_lake_min = 0;

def lake(dom):
	x = dom[0];
	y = dom[1];
	z = z_lake+0.8;
 	return [x, y, z];

def zLake(dom):
	x = dom[0];
	y = dom[1];
	z=0.8;
	return [x, y, z];

MAP(zLake)(domainLake);
l = T([1,2])([20,20])(COLOR(BLUE)(MAP(lake)(domainLake)));
s = STRUCT([terrain,l,terrain]);
VIEW(s);

File: 2013-06-07/javascript/exercise2.js
var domainMounts = DOMAIN([[-50, 50], [-50, 50]])([32, 32]);
var altitude = {};

var mounts = function(dom){
    var x = dom[0];
    var y = dom[1];
    var z = Math.random()*Math.abs(x%y);
    if (z==0)
    	z = Math.random();
    this.altitude[''+Math.floor(x)+','+Math.floor(y)] = z;
 return [x, y, z];
}
var terrain = MAP(mounts)(domainMounts);

var domainLake = DOMAIN([[-10, 10], [-10, 10]])([32, 32]);

var z_lake = 0;
var z_lake_min = 0;

var lake = function(dom){
    var x = dom[0];
    var y = dom[1];
 	 var z = this.z_lake+0.8;
    return [x, y, z];
}

var zLake = function(dom){
    var x = dom[0];
    var y = dom[1];
 	 var z = this.altitude[''+Math.floor(x)+','+Math.floor(y)];
 	 if (z < this.z_lake)
 	 	this.z_lake = z;
 	 return [x, y, z];
}
MAP(zLake)(domainLake);
var l = COLOR([139/255,232/255,218/255])(MAP(lake)(domainLake));
var s = STRUCT([terrain,l]);
DRAW(s);